process_duty_email.py

```diff
@@ -0,0 +1,293 @@
+"""
+Process duty emails
+Developer: Igor Malashchuk 
+Email: igor.malashchuk@nhs.net
+Date Modified: 14/09/2022
+"""
+import subprocess
+import os
+import re
+import collections
+from datetime import datetime
+import pandas
+import math
+import numpy as np
+import config
+
+# Run script in Powershell: duty
+# run script in command prompt: S:\Genetics_Data2\Array\Software\Python-3.6.5\python S:\Genetics_Data2\Array\Software\duty_bioinformatician_scripts\process_duty_email.py
+version = "2.0.0"
+
+def ask_for_folder():
+    """
+    For MokaPipe the destination folder can be different.
+    Therefore this function ask the user to enter the last part of the folder
+    """
+    output_folder = ""
+    path_worksheets = config.MokaPipe_path_worksheets
+    print("Part of destination is:{}".format(path_worksheets))
+    while len(output_folder) == 0:
+        output_folder = input(
+            'Please complete the final part of the destination folder \n(e.g."NGS_401 to 500\\NGS484")\nMay need to place inside double quotes in Powershell:'
+        )
+        destination_folder = path_worksheets + output_folder
+        print("The destination folder is : {}".format(destination_folder))
+        if destination_folder == path_worksheets or not os.path.isdir(destination_folder):
+            print("The directory does not exist. Please try again!")
+            output_folder = ""
+    return path_worksheets + output_folder
+
+def get_data(df, path_to_folder):
+    """
+    This function reads the data frame (df)..
+    From the df file it obtains the download links and creates a string argument.   
+    """
+    all_urls = ""
+    for index, row in df.iterrows():
+        line = row['url'] + "," + path_to_folder + "£$%"
+        all_urls += line
+    return all_urls
+
+def download_data(all_urls):
+    """"
+    This function checks if the string argument is less than the limit. 
+    If it is longer than the character length limit thee string is split into multiple smaller string arguments. 
+    The function then downloads the data using powershell.
+    """
+    limit = 8000
+    result = math.ceil(len(all_urls)/limit)
+    if result <= 1:
+        urls_array = all_urls.split("£$%")
+        merged_array = " ".join(urls_array)
+        urls_list = [merged_array]
+    else:
+        urls_array = all_urls.split("£$%")

```

Comment:
could you just split the all_urls regardless if it's > 8000 characters?

Reply:
eg just fire off a powershell cmd for each file?

Reply:
I found that when I create a loop of many powershell instances often one of them would get an error. Therefore, I tried to reduce as much as possible the number of PowerShell requests by providing as many URL files as possible in one string.